### ISSUE TITLE
Generalized VkStruct?

### ISSUE THREAD

**Issue by Rotaerk** [achirkin/vulkan#45]
Have you considered, or do you see any problem with, the idea of making a generalization of VkStruct, one that's not Vulkan specific?  It just seems to me that the ability to read/write a byte array as a struct with named and typed fields would be useful elsewhere.  Particularly in C-library bindings.

**Comment by achirkin** [comment]
The funny thing is that this is exactly what [`PrimBytes`](http://hackage.haskell.org/package/easytensor-2.1.0.0/docs/Numeric-PrimBytes.html#t:PrimBytes) achieves :) I did not aim at this, but it arose from [our discussion](https://github.com/achirkin/easytensor/issues/13) of how to get field offsets for a user-defined `Vertex` type. I was thinking whether I should split PrimBytes from easytensor though.

**Comment by Rotaerk** [comment]
Ah, nice! I do think breaking it off into its own library would be useful.  It could be used by both easytensor and vulkan-api (if you were willing to give it a dependency).

But I'm looking to make or modify a binding to assimp, and I noticed the existing one is defining simple record types and Storable instances.  I realized it might be nicer to define VkStruct like things instead...

**Comment by Rotaerk** [comment]
Though, whatever was extracted into its own library would need to support arrays of structs, which is what your dataframes fundamentally are...  so maybe you should have a generalized primbytes+dataframe library, and then just put tensor-specific stuff in easytensor?

**Comment by achirkin** [comment]
Yes, this would make sense! Except there is no tensor-specific stuff in `easytensor` yet :-D
I consider `PrimBytes` more as just an alternative to `Storable` and thinking of packing it as such. But I am not sure that many people will use it outside easytensor/vulkan-api anywhere, so I don't hurry with this decision so far.

**Comment by Rotaerk** [comment]
Well the tensor-specific stuff is your Matrix and Vertex stuff, for example.  DataFrame is more general than being used as a tensor, since it can be treated as just a typesafe array of data.  At least that's how I see it currently.

**Comment by Rotaerk** [comment]
And regarding people using it outside vulkan/easytensor, it seems like it would be very useful for FFI libraries, and code using them.  Besides that, I can't think of why you would need it...  If you don't need a pointer to precisely laid out data, you can just use standard ADTs.

**Comment by achirkin** [comment]
NB: there is some duplicated functionality in `PrimBytes` and `VkStruct`. If we were to split `PrimBytes` from `easytensor`, it's worth reusing it here.

https://github.com/achirkin/easytensor/projects/2
